lib/sprockets/utils.rb

```diff
@@ -97,11 +100,21 @@ def string_end_with_semicolon?(str)
     #
     # Returns buf String.
     def concat_javascript_sources(buf, source)
-      if buf.bytesize > 0
-        buf << ";" unless string_end_with_semicolon?(buf)
-        buf << "\n" unless buf.end_with?("\n")
+      if source.bytesize > 0
+        buf << source
+
+        # If the source contains non-ASCII characters, indexing on it becomes O(N).
+        # This will lead to O(N^2) performance in string_end_with_semicolon?, so we should use 32 bit encoding to make sure indexing stays O(1)
+        source = source.encode(Encoding::UTF_32LE) unless source.ascii_only?
```

Comment:
I thought about that and have a commit that implements this here: https://github.com/bouk/sprockets/commit/9d3edc991fe2e2786b99f1eaa789be70c2bcb67c

The issue is that it needs to always use UTF32 for the buffer, which leads to up to 4x the memory usage and slower performance